framework/src/service/ServiceListeners.cpp

```diff
@@ -122,48 +155,94 @@ void ServiceListeners::RemoveServiceListener(const ServiceListenerEntry& entryTo
+
   auto l = bundleListenerMap.Lock(); US_UNUSED(l);
   auto& listeners = bundleListenerMap.value[context];
-  if (std::find_if(listeners.begin(), listeners.end(), std::bind(BundleListenerCompare(), std::make_pair(listener, data), std::placeholders::_1)) == listeners.end())
-  {
-    listeners.push_back(std::make_pair(listener, data));
-  }
+  listeners[token.Id()] = std::make_tuple(listener, data);
+  return token;
 }
 
+/**
+ * Called by the deprecated RemoveBundleListener(name_of_callable)
+ */
 void ServiceListeners::RemoveBundleListener(const std::shared_ptr<BundleContextPrivate>& context, const BundleListener& listener, void* data)
 {
   auto l = bundleListenerMap.Lock(); US_UNUSED(l);
-  bundleListenerMap.value[context].remove_if(std::bind(BundleListenerCompare(), std::make_pair(listener, data), std::placeholders::_1));
+  auto& listeners = bundleListenerMap.value[context];
+  auto it = std::find_if(listeners.begin(), listeners.end(), std::bind(BundleListenerCompareListenerData, listener, data, std::placeholders::_1));
+  if (it != listeners.end())
+  {
+      listeners.erase(it);
+  }
 }
 
-void ServiceListeners::AddFrameworkListener(const std::shared_ptr<BundleContextPrivate>& context, const FrameworkListener& listener, void* data)
+ListenerToken ServiceListeners::AddFrameworkListener(const std::shared_ptr<BundleContextPrivate>& context, const FrameworkListener& listener, void* data)
 {
+  auto token = MakeListenerToken();
+
   auto l = frameworkListenerMap.Lock(); US_UNUSED(l);
   auto& listeners = frameworkListenerMap.value[context];
-  if (std::find_if(listeners.begin(), listeners.end(), std::bind(FrameworkListenerCompare(), std::make_pair(listener, data), std::placeholders::_1)) == listeners.end())
-  {
-    listeners.push_back(std::make_pair(listener, data));
-  }
+  listeners[token.Id()] = std::make_tuple(listener, data);
+  return token;
 }
 
+/**
+ * Called by the deprecated RemoveFrameworkListener(name_of_callable)
+ */
 void ServiceListeners::RemoveFrameworkListener(const std::shared_ptr<BundleContextPrivate>& context, const FrameworkListener& listener, void* data)
 {
   auto l = frameworkListenerMap.Lock(); US_UNUSED(l);
   auto& listeners = frameworkListenerMap.value[context];
-  listeners.erase(std::remove_if(listeners.begin(),
-                                 listeners.end(),
-                                 std::bind(FrameworkListenerCompare(), std::make_pair(listener, data), std::placeholders::_1)),
-                  listeners.end());
+  auto it = std::find_if(listeners.begin(), listeners.end(), std::bind(FrameworkListenerCompareListenerData, listener, data, std::placeholders::_1));
+  if (it != listeners.end())
+  {
+    listeners.erase(it);
+  }
+}
+
+void ServiceListeners::RemoveListener(const std::shared_ptr<BundleContextPrivate>& context, ListenerToken token)
+{
+  if (!token)
+  {
+    return;
+  }
+
+  auto tokenId = token.Id();
+  {
+    auto l = frameworkListenerMap.Lock(); US_UNUSED(l);
```

Comment:
This block and the folowing block have the same functionality. can we create one function/lambda and call it twice instead of duplicating code with different variables.